### PR TITLE
Fix 'IllegalStateException is thrown when executing in HALF_OPEN state'

### LINKED ISSUE #41: Future's IllegalStateException is thrown when executing in HALF_OPEN state
Version: 3.8.4

### Context

I use a CircuiBreaker to talk to a backend, so I want to probe its health whenever the breaker becomes half-open to give it an opportunity to go back to closed, because when the breaker is open that backend is marked as unhealthy and never selected to serve a request.
I set a timeout on my circuit breaker.
Whenever the half-open handler executes on the breaker and the execution completes (either success/fail/timeout), an `IllegalStateException` is logged:

```
SEVERE: Unhandled exception
java.lang.IllegalStateException: Result is already complete: failed
	at io.vertx.core.impl.FutureImpl.fail(FutureImpl.java:128)
	at io.vertx.circuitbreaker.impl.CircuitBreakerImpl.lambda$null$4(CircuitBreakerImpl.java:225)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:369)
	at io.vertx.core.impl.EventLoopContext.lambda$executeAsync$0(EventLoopContext.java:38)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:510)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:518)
	at io.netty.util.concurrent.SingleThreadEventExecutor$6.run(SingleThreadEventExecutor.java:1044)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

### Do you have a reproducer?

https://gist.github.com/Feder1co5oave/8d70b3383fd6c2011619f38e17d19d46#file-cbtest-java

Step by step:
- breaker is closed
- execute with success
- breaker remains closed
- execute with fail (or timeout, here the behavior is correct)
- breaker becomes open
- wait for it to become half-closed
- execute with timeout
- breaker becomes open and IllegalStateException is thrown from CircuitBreakerImpl.

### Extra
JVM is OpenJDK 11 on Ubuntu Linux.

### Fix

This is because when the breaker is in the `HALF_OPEN` state, **two** handlers are added to the resultFuture in [`executeAndReportWithFallback`](https://github.com/vert-x3/vertx-circuit-breaker/blob/db67d3d16144344609077b659c0342026a674806/src/main/java/io/vertx/circuitbreaker/impl/CircuitBreakerImpl.java#L200), and both take care of updating the breaker's state and metrics. Only the second handler should be set when the breaker is HALF_OPEN.
Additionally, I think a `runOnContext` should be added to the handler in the HALF_OPEN case, to ensure that the result handler is run on the same context as the caller, similarly to what was done in 25ac02cdf.

### PR DESCRIPTION
**Pull request by Feder1co5oave**
Fixes #41

Conformance:
- [x] sign the ECA

### CODE REVIEW

File: src/test/java/io/vertx/circuitbreaker/impl/CircuitBreakerMetricsTest.java [hunk 1]
```diff
@@ -215,7 +215,7 @@ public void testLatencyComputation(TestContext tc) {
   }
 
   @Test
-  @Repeat(100)
+  @Repeat(50)
```
Comment (cescoffier) [inline src/test/java/io/vertx/circuitbreaker/impl/CircuitBreakerMetricsTest.java]: Any reason to reduce the number of iteration?

**Comment by vietj** [comment]
thanks for the review @cescoffier

File: src/test/java/io/vertx/circuitbreaker/impl/CircuitBreakerMetricsTest.java (hunk 1 above)
Comment (Feder1co5oave) [reply src/test/java/io/vertx/circuitbreaker/impl/CircuitBreakerMetricsTest.java]: Yes, I noticed that after having added my test unit and run the test with `mvn test`, it failed with a StackOverflowException with a stack almost entirely made out of composite Futures. I took a hunch and guessed that it was caused by the high number of iterations. I re-ran the tests now and terminated successfully. I'm going to revert this commit!

Comment (Feder1co5oave) [reply src/test/java/io/vertx/circuitbreaker/impl/CircuitBreakerMetricsTest.java]: edit: re-ran the tests **with the original number of iterations**

**Comment by srijan02420** [comment]
@Feder1co5oave @cescoffier when is it going to get merged?

**Comment by srijan02420** [comment]
@vietj please release new version with latest master.

**Comment by vietj** [comment]
@srijan02420 this will be in Vert.x 3.9.0 expected early march

**Pull request merged**